### PR TITLE
feat(api): implement branch/commit comparison API

### PR DESCRIPTION
**Pull request by appleboy**
- Add new `Compare` struct to represent comparison between two commits
- Introduce new API endpoint `/compare/*` to get commit comparison information
- Create new file `repo_compare.go` with the `Compare` struct definition
- Add new file `compare.go` in `routers/api/v1/repo` to handle comparison logic
- Add new file `compare.go` in `routers/common` to define `CompareInfo` struct
- Refactor `ParseCompareInfo` function to use `common.CompareInfo` struct
- Update Swagger documentation to include the new API endpoint for commit comparison
- Remove duplicate `CompareInfo` struct from `routers/web/repo/compare.go`
- Adjust base path in Swagger template to be relative (`/api/v1`)

GitHub API https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits

### CODE REVIEW

**Comment by lunny** [comment]
Sorry, the code looks like duplicated to previous blocks.

**Comment by appleboy** [comment]
@lunny Yes, so I remove some unused logic for API

**Comment by lunny** [comment]
It's better to have an integration test for the API.

**Comment by appleboy** [comment]
@lunny See the testing https://github.com/go-gitea/gitea/pull/30349/files#diff-41e2d9c42fe9e4bf3e8eb44f89eef65b5bc77dfb19c4692092686b4bcf5f5151

**Comment by appleboy** [comment]
Added permission check [`6ce81db` (#30349)](https://github.com/go-gitea/gitea/pull/30349/commits/6ce81dba1b1c5eb716ee08c7557e2984a29bcc72)

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -1066,6 +1066,8 @@ func Routes() *web.Route {
 			m.Post("/migrate", reqToken(), bind(api.MigrateRepoOptions{}), repo.Migrate)
 
 			m.Group("/{username}/{reponame}", func() {
+				m.Get("/compare/*", reqAnyRepoReader(), repo.CompareDiff)
```
Comment (lunny) [inline routers/api/v1/api.go]:
```suggestion
				m.Get("/compare/*", reqRepoReader(unit.TypeCode), repo.CompareDiff)
```

**Review by lunny — APPROVED** [review]

**Comment by wolfogre** [comment]
Sorry, I cannot give my approval since the `ParseCompareInfo` function is almost copied from https://github.com/go-gitea/gitea/blob/708e87e17df2b6a03eca3cac026a51beed296b5b/routers/web/repo/compare.go#L200-L574

It contains many logics and detailed comments which make me believe it contains complex logic and shouldn't be maintained in two places.

**Comment by appleboy** [comment]
@wolfogre Yes, in the next phase I will refactor the `ParseCompareInfo` from route/web and route/api.

**Comment by appleboy** [comment]
@wolfogre  @lunny  See the latest commit. https://github.com/go-gitea/gitea/pull/30349/commits/94746ecc56147f2265960d0c140cc314137f74df

**Review by lunny — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

**Pull request merged**